Source/Core/Core/AchievementManager.cpp

```diff
@@ -262,6 +266,129 @@ bool AchievementManager::IsHardcoreModeActive() const
   return rc_client_is_processing_required(m_client);
 }
 
+bool AchievementManager::VerifyPatchHash(const Common::SHA1::Digest& digest) const
+{
+  std::string sys_path = File::GetSysDirectory() + DIR_SEP;
+
+  File::IOFile list_file;
+  list_file.Open(fmt::format("{}{}", sys_path, APPROVED_LIST_FILENAME), "rb");
+  std::vector<u8> buffer(list_file.GetSize());
+  list_file.ReadBytes(buffer.data(), list_file.GetSize());
+  list_file.Close();
+
+  auto context = Common::SHA1::CreateContext();
+  context->Update(buffer);
+  auto list_digest = context->Finish();
+  std::string list_hash;
+  for (size_t ix = 0; ix < list_digest.size(); ix++)
+  {
+    u8 upper = list_digest[ix] / 16;
+    list_hash.append(1, upper + ((upper > 9) ? 'A' - 10 : '0'));
+    u8 lower = list_digest[ix] % 16;
+    list_hash.append(1, lower + ((lower > 9) ? 'A' - 10 : '0'));
+  }
```

Comment:
Please do this the other way around. Instead of converting the hash to hex, just store the hash in the correct binary representation.

If you need a function that generates the binary representation from the string, you can use this one: https://github.com/AdmiralCurtiss/SenPatcher/blob/7c6e406ebc4187f1f08332713eb88f00173c5a5d/native/util/hash/sha1.h#L16-L28

Reply:
I'm not sold on this change, or I'm not understanding what you're asking. The hashes will be saved in a text file alongside comments, and if a hash doesn't match I want to log it so that I can easily copy the hash from the logs and insert it into the file. This gets a lot hairier if the hash might have special characters in it; if the characters aren't meant to be visible I can't even tell them apart easily. I wouldn't be opposed to putting this kind of conversion in a utility somewhere though if that would help?

Reply:
I mean that you take the string, convert it to the actual 20-byte hash, and then compare the hash, instead of comparing the strings. But this isn't critical, we can improve this later if you for some reason don't want to do it now.

Reply:
My point was that, while I can do that, I need to report the result in a way that can be updated in the hash file, which means instead of just converting the actual hash to a string and comparing it with the expected string, I'm converting the expected string to a hash, comparing it, and then if there's an error converting the actual hash to a string anyways. Why not save the extra compare?